frontend/controller/app/identity.js

```diff
@@ -0,0 +1,252 @@
+    try {
+      const userID = await sbp('gi.actions/identity/create', {
+        // TODO: Wrap IPK and IEK in "Secret"
+        IPK: serializeKey(IPK, true),
+        IEK: serializeKey(IEK, true),
+        publishOptions,
+        username,
+        email,
+        picture,
+        r,
+        s,
+        sig,
+        Eh
+      })
+
+      return userID
+    } catch (e) {
+      console.error('gi.app/identity/create failed!', e)
+      throw new GIErrorUIRuntimeError(L('Failed to create user identity: {reportError}', LError(e)))
+    }
+  },
+  'gi.app/identity/signup': async function ({ username, email, password }, publishOptions) {
+    try {
+      const randomAvatar = sbp('gi.utils/avatar/create')
+      const userID = await sbp('gi.app/identity/create', {
+        data: {
+          username,
+          email,
+          password,
+          picture: randomAvatar
+        },
+        publishOptions
+      })
+      return userID
+    } catch (e) {
+      console.error('gi.app/identity/signup failed!', e)
+      await sbp('gi.app/identity/logout') // TODO: should this be here?
+      const message = LError(e)
+      if (e.name === 'GIErrorUIRuntimeError') {
+        // 'gi.app/identity/create' also sets reportError
+        message.reportError = e.message
+      }
+      throw new GIErrorUIRuntimeError(L('Failed to signup: {reportError}', message))
+    }
+  },
+  'gi.app/identity/login': async function ({ username, password: wpassword, identityContractID }: {
+    username: ?string, password: ?Secret<string>, identityContractID: ?string
+  }) {
+    if (username) {
+      identityContractID = await sbp('namespace/lookup', username)
+    }
+
+    if (!identityContractID) {
+      throw new GIErrorUIRuntimeError(L('Incorrect username or password'))
+    }
+
+    const password = wpassword?.valueOf()
+    const transientSecretKeys = []
+    if (password) {
+      try {
+        const salt = await sbp('gi.app/identity/retrieveSalt', username, wpassword)
+        const IEK = await deriveKeyFromPassword(CURVE25519XSALSA20POLY1305, password, salt)
+        transientSecretKeys.push(IEK)
+      } catch (e) {
+        console.error('caught error calling retrieveSalt:', e)
+        throw new GIErrorUIRuntimeError(L('Incorrect username or password'))
+      }
+    }
+
+    try {
+      sbp('appLogs/startCapture', identityContractID)
+      const { state, cheloniaState, encryptionParams } = await loadState(identityContractID, password)
+      let loginCompleteHandler, loginErrorHandler
+
+      try {
+        const loginCompletePromise = new Promise((resolve, reject) => {
+          const loginCompleteHandler = ({ identityContractID: id }) => {
+            sbp('okTurtles.events/off', LOGIN_ERROR, loginErrorHandler)
+            if (id === identityContractID) {
+              // We need to save the state to ensure it's possible to refresh
+              // the page
+              resolve(sbp('state/vuex/save'))
```

Comment:
Passing in `sbp('state/vuex/save')` to the resolved promise doesn't seem to do anything... so either a `// TODO` comment is needed or this needs to be rethought. Th

Reply:
Yes, this does something. `state/vuex/save` is an async function, so it'll make the promise continue to wait until that resolves.

Reply:
OK, could you add a comment stating that and also explaining why you're doing this?

Reply:
There already is a comment, right in the part you quoted.

Reply:
Are you referring to this?

```js
              // We need to save the state to ensure it's possible to refresh
              // the page
```

If so, that comment doesn't help me understanding why `sbp('state/vuex/save')` is being called here.

The line below (314) does nothing with this value:

```js
await loginCompletePromise
```

Reply:
save the state => sbp('state/vuex/save')
The line there does something with that value, namely waiting.

Reply:
OK, sorry, the problem here was that my brain was being momentarily dyslexic or something, I think. It was so used to reading `'state/vuex/state'`, that I didn't see this was a totally different selector called `'state/vuex/save'`. I thought you were passing in `'state/vuex/state'`.